### PR TITLE
Feature/cicd

### PR DESCRIPTION
**Pull request by sridharvikram**
I have added cicd pipeline code and readme doc to set it up for migrating agent under examples folder. I would like contribute this to SCRAPI library. Please review the changes

### CODE REVIEW

File: examples/dfcx_agent_cicd/cicd_code/export/export.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+[3] Add Version to ready to deploy Environment (in dfcx)
+
+Handled by the pipline
+[4] Run the pipeline to Export ready to deploy Environment / Agent
+[5] Export metadata file to gcs in path same as agent
+"""
+
+
+
+def agentToGcs(
```
Comment (kmaphoenix) [inline examples/dfcx_agent_cicd/cicd_code/export/export.py:30]: Need to follow snake_case python naming standards.

File: examples/dfcx_agent_cicd/cicd_code/export/export.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    logging.info("Agent export from dev done")
+
+def metaToGcs(
+    config,
+    source_flow_names,
+    updatedcommitmessage,
+    gs_loc,
+    source_agent_name,
+    bucket,
+    impactedVersionIds):
```
Comment (kmaphoenix) [inline examples/dfcx_agent_cicd/cicd_code/export/export.py:58]: prefer snake_case

File: examples/dfcx_agent_cicd/cicd_code/export/flowimpacted.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
+from dfcx_scrapi.core.agents import Agents
+from dfcx_scrapi.core.versions import Versions
+from dfcx_scrapi.core.environments import Environments
+from dfcx_scrapi.core.flows import Flows
+
+from typing import Dict, List
+
+class impacted:
```
Comment (kmaphoenix) [inline examples/dfcx_agent_cicd/cicd_code/export/flowimpacted.py:8]: class names should be capitalized in Python.
2 spaces before the class starts

File: examples/dfcx_agent_cicd/cicd_code/export/flowimpacted.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+            flow=ver.split("/")[-3]
+            if flow in impactedflows:
+                impactedVersionIds.append(ver)
+
+
+        return filtered_dict,impactedVersionIds
+
+
+
+    def checkFlow(
```
Comment (kmaphoenix) [inline examples/dfcx_agent_cicd/cicd_code/export/flowimpacted.py:68]: snake_case

File: examples/dfcx_agent_cicd/cicd_code/readme.md [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+    lines about your change in the agent that will be used )
+
+3.  This will export agent and this would have done a check in in the
+    repo to trigger UAT and prod builds and deployed the agent in UAT
+    project.
+
+4.  Now you can come back to cloud build console and build history tab
+    and approve the build that is waiting for your approval and you can
+    see that it will deploy the agent in prod post approval
+    !media/image6.png
```
Comment (kmaphoenix) [inline examples/dfcx_agent_cicd/cicd_code/readme.md:306]: this image is not loading in the markdown file

File: examples/dfcx_agent_cicd/cicd_code/readme.md [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+
+## Cavet
```
Comment (kmaphoenix) [inline examples/dfcx_agent_cicd/cicd_code/readme.md:308]: sp: `Caveat`

File: examples/dfcx_agent_cicd/cicd_code/shared/deployment.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+[7] sync the deployed version back deployed env in dev
+"""
+
+class Deployment:
+    def __init__(self,input_dict):
+        for key, value in input_dict.items():
+            setattr(self, key, value)     
+
+    #[1] import to destination project
+    def importAgent(self,webhookenv):
```
Comment (kmaphoenix) [inline examples/dfcx_agent_cicd/cicd_code/shared/deployment.py:45]: prefer snake_case

File: examples/dfcx_agent_cicd/cicd_code/shared/enVsOtherLang.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+      client = pages.PagesClient(
+          credentials=self.creds, client_options=client_options
+      )
+
+      response = client.get_page(request)
+
+      return response
+
+
+def enVslang(agent_id,flows,lang):
```
Comment (kmaphoenix) [inline examples/dfcx_agent_cicd/cicd_code/shared/enVsOtherLang.py:42]: prefer snake_case

File: examples/dfcx_agent_cicd/cicd_code/shared/fullfillmentHelper.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
+
+
+def getEntryff(page=None,page_id=None,language_code='en'):
```
Comment (kmaphoenix) [inline examples/dfcx_agent_cicd/cicd_code/shared/fullfillmentHelper.py:3]: prefer snake_case

File: examples/dfcx_agent_cicd/cicd_code/shared/testcaserun.py [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+import logging
+
+# logging config
+logging.basicConfig(
+    level=logging.INFO,
+    format="%(asctime)s %(levelname)-8s %(message)s",
+    datefmt="%Y-%m-%d %H:%M:%S",
+)
+
+class runtestcases:
```
Comment (kmaphoenix) [inline examples/dfcx_agent_cicd/cicd_code/shared/testcaserun.py:16]: capitalize Class names

File: examples/dfcx_agent_cicd/cicd_code/shared/testcaserun.py [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+
+        #get environment id
+        if self.environment_name:
+            env_details=env.get_environment_by_display_name(display_name=self.environment_name
+                                                            ,agent_id=self.agent_id)
+            self.env_id=env_details.name
+        else:
+            self.env_id=None
+
+    def triggerTestcase(
```
Comment (kmaphoenix) [inline examples/dfcx_agent_cicd/cicd_code/shared/testcaserun.py:43]: snake_case

**Review by kmaphoenix — CHANGES_REQUESTED** [review]
Hi @sridharvikram there are several changes I've tagged in the files that are Python style updates that need to be made.

In addition, please run `pylint` locally to see all of the linter errors that need to be corrected.
You can also view the latest pylint run directly in the repo here: 
https://github.com/GoogleCloudPlatform/dfcx-scrapi/actions/runs/9990092140/job/27632420520?pr=196

File: examples/dfcx_agent_cicd/cicd_code/export/export.py (hunk 1 above)
Comment (sridharvikram) [reply examples/dfcx_agent_cicd/cicd_code/export/export.py]: Fixed in my latest push to my branch

File: examples/dfcx_agent_cicd/cicd_code/export/flowimpacted.py (hunk 3 above)
Comment (sridharvikram) [reply examples/dfcx_agent_cicd/cicd_code/export/flowimpacted.py]: Fixed in my latest push to my branch

File: examples/dfcx_agent_cicd/cicd_code/readme.md (hunk 5 above)
Comment (sridharvikram) [reply examples/dfcx_agent_cicd/cicd_code/readme.md]: Fixed in my latest push to my branch

File: examples/dfcx_agent_cicd/cicd_code/shared/deployment.py (hunk 7 above)
Comment (sridharvikram) [reply examples/dfcx_agent_cicd/cicd_code/shared/deployment.py]: Fixed in my latest push to my branch

File: examples/dfcx_agent_cicd/cicd_code/shared/fullfillmentHelper.py (hunk 9 above)
Comment (sridharvikram) [reply examples/dfcx_agent_cicd/cicd_code/shared/fullfillmentHelper.py]: Fixed in my latest push to my branch

File: examples/dfcx_agent_cicd/cicd_code/shared/testcaserun.py (hunk 10 above)
Comment (sridharvikram) [reply examples/dfcx_agent_cicd/cicd_code/shared/testcaserun.py]: Fixed in my latest push to my branch

File: examples/dfcx_agent_cicd/cicd_code/shared/testcaserun.py (hunk 11 above)
Comment (sridharvikram) [reply examples/dfcx_agent_cicd/cicd_code/shared/testcaserun.py]: Fixed in my latest push to my branch

File: examples/dfcx_agent_cicd/cicd_code/shared/enVsOtherLang.py (hunk 8 above)
Comment (sridharvikram) [reply examples/dfcx_agent_cicd/cicd_code/shared/enVsOtherLang.py]: Fixed in my latest push to my branch

File: examples/dfcx_agent_cicd/cicd_code/readme.md (hunk 6 above)
Comment (sridharvikram) [reply examples/dfcx_agent_cicd/cicd_code/readme.md]: Fixed in my latest push to my branch

File: examples/dfcx_agent_cicd/cicd_code/export/export.py (hunk 2 above)
Comment (sridharvikram) [reply examples/dfcx_agent_cicd/cicd_code/export/export.py]: Fixed in my latest push to my branch

File: examples/dfcx_agent_cicd/cicd_code/export/flowimpacted.py (hunk 4 above)
Comment (sridharvikram) [reply examples/dfcx_agent_cicd/cicd_code/export/flowimpacted.py]: Fixed in my latest push to my branch

**Comment by sridharvikram** [comment]
@kmaphoenix I have addressed all the lint issues you have mentioned and also fixed most of the lint errors that i ran in my IDE. Please re review it and do the needful.

**Comment by kmaphoenix** [comment]
Hi @sridharvikram , did you perform a `git rebase` with `main` branch?

It's showing many edits to files that are not related to your CICD code and should not have been modified. 

If you did not perform a rebase, you should do this first. 
If you're not familiar with running a rebase, you can consider starting a new branch from the latest main branch. 

Otherwise it makes it very difficult to review all of your code.

**Comment by kmaphoenix** [comment]
@sridharvikram yeah after looking at the branch some more there are many merge conflicts. 

Best course of action:
- Start a new branch from `main`
- Add your changes for CICD code
- Submit PR

Otherwise we're going to have a lot of back and forth to get this fixed and you'll need to fix all of the merge conflicts locally.

**Comment by sridharvikram** [comment]
@kmaphoenix I have created this new PR https://github.com/GoogleCloudPlatform/dfcx-scrapi/pull/239 as suggested. Hence closing this one. Thanks

**Pull request closed without merging**